### PR TITLE
CredentialsEndpointProvider.java: Adds the sending of a User-Agent other than default User-Agent to the metadata service

### PR DESCRIPTION
**Pull request by willbengtson**
Adds the sending of a User-Agent other than default User-Agent in Java when making requests to metadata service

Similar to AWS Golang SDK and Ruby SDK, send a User-Agent other than the default User-Agent when making requests to the EC2 metadata for credentials.
https://github.com/aws/aws-sdk-java/pull/1562 for aws-sdk-java addresses this also.  PR #1445 for botocore implements this too.

This will enable protection of AWS credentials from Server Side Request Forgery (SSRF) vectors by being able to use a metadata proxy to block requests that do not have the right User-Agent set.  User-Agent is not controllable by an attacker via SSRF.

## Description
Set default headers in the `CredentialsEndpointProvider` interface

## Motivation and Context
This will enable protection of AWS credentials from Server Side Request Forgery (SSRF) vectors by being able to use a metadata proxy to block requests that do not have the right User-Agent set.  User-Agent is not controllable by an attacker via SSRF.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `mvn install` succeeds
- [x] My code follows the code style of this project
- [ ] My change requires a change to the Javadoc documentation
- [ ] I have updated the Javadoc documentation accordingly
- [x] I have read the **README** document
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed
- [x] A short description of the change has been added to the **CHANGELOG**

## License

- [x] I confirm that this pull request can be released under the Apache 2 license

### CODE REVIEW

File: core/src/main/java/software/amazon/awssdk/core/internal/CredentialsEndpointProvider.java [hunk 1]
```diff
@@ -56,7 +57,12 @@ default CredentialsEndpointRetryPolicy retryPolicy() {
      * Allows passing additional headers to the request
      */
     default Map<String, String> headers() {
-        return Collections.emptyMap();
+        Map<String, String> requestHeaders = new HashMap<String, String>();
```
Comment (zoewangg) [inline core/src/main/java/software/amazon/awssdk/core/internal/CredentialsEndpointProvider.java]: nit: can we use diamond operator here?

**Comment by willbengtson** [comment]
@zoewangg the tests fail because of NullPointerException.  I'm trying to use `import software.amazon.awssdk.core.util.VersionInfo;` but for some reason it's should Null when running the tests. I'm not super familiar with what to modify to get the test to pick the file up correctly.  Am I seeing that correctly?

**Comment by zoewangg** [comment]
Once the build succeeds, could you add a new-change entry? you can do that by running the scripts/new-change script and following the instructions.

Can you confirm that this PR is submitted under Apache 2.0 license by checking the ` I confirm that this pull request can be released under the Apache 2 license` box under license section in the description?

File: core/src/main/java/software/amazon/awssdk/core/internal/CredentialsEndpointProvider.java (hunk 1 above)
Comment (willbengtson) [reply core/src/main/java/software/amazon/awssdk/core/internal/CredentialsEndpointProvider.java]: sure thing

**Comment by willbengtson** [comment]
Will do, working on these changes

**Comment by willbengtson** [comment]
@zoewangg updated

**Comment by zoewangg** [comment]
@willbengtson Thanks for making the changes! LGTM. 

Looks like the branch is out-of-date with master. Could you do a merge from master and then squash to a single commit? I'll merge it afterwards.

**Comment by willbengtson** [comment]
@zoewangg Done

**Pull request merged**